hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/LatencyTracker.java

```diff
@@ -0,0 +1,215 @@
+
+    recordClientLatency(operationStart, operationStop, callerName, calleeName, success, res);
+  }
+
+  public void recordClientLatency(
+          Instant operationStart,
+          Instant operationStop,
+          String callerName,
+          String calleeName,
+          boolean success,
+          AbfsHttpOperation res) {
+    if (!enabled) {
+      return;
+    }
+
+    Instant trackerStart = Instant.now();
+    long latency = isValidInstant(operationStart) && isValidInstant(operationStop)
+            ? Duration.between(operationStart, operationStop).toMillis() : -1;
+
+    String latencyDetails = String.format(singletonLatencyReportingFormat,
+            Instant.now(),
+            callerName,
+            calleeName,
+            success ? "Succeeded" : "Failed",
+            latency,
+            res == null ? "" : (" " + res.toKvpString()));
+
+    this.offerToQueue(trackerStart, latencyDetails);
+  }
+
+  public void recordClientLatency(
+          Instant operationStart,
+          String callerName,
+          String calleeName,
+          boolean success,
+          Instant aggregateStart,
+          long aggregateCount,
+          AbfsHttpOperation res) {
+    if (!enabled) {
+      return;
+    }
+
+    Instant operationStop = getLatencyInstant();
+
+    recordClientLatency(operationStart, operationStop, callerName, calleeName, success, aggregateStart, aggregateCount, res);
+  }
+
+  public void recordClientLatency(
+          Instant operationStart,
+          Instant operationStop,
+          String callerName,
+          String calleeName,
+          boolean success,
+          Instant aggregateStart,
+          long aggregateCount,
+          AbfsHttpOperation res){
+    if (!enabled) {
+      return;
+    }
+
+    Instant trackerStart = Instant.now();
+    long latency = isValidInstant(operationStart) && isValidInstant(operationStop)
+            ? Duration.between(operationStart, operationStop).toMillis() : -1;
+    long aggregateLatency = isValidInstant(aggregateStart) && isValidInstant(operationStop)
+            ? Duration.between(aggregateStart, operationStop).toMillis() : -1;
+
+    String latencyDetails = String.format(aggregateLatencyReportingFormat,
+            Instant.now(),
+            callerName,
+            calleeName,
+            success ? "Succeeded" : "Failed",
+            latency,
+            aggregateLatency,
+            aggregateCount,
+            res == null ? "" : (" " + res.toKvpString()));
+
+    offerToQueue(trackerStart, latencyDetails);
+  }
+
+  public String getClientLatency() {
+    if (!enabled) {
+      return null;
+    }
+
+    Instant trackerStart = Instant.now();
+    String latencyDetails = queue.poll(); // non-blocking pop
+
+    if (LOG.isDebugEnabled()) {
+      Instant stop = Instant.now();
+      long elapsed = Duration.between(trackerStart, stop).toMillis();
+      LOG.debug(String.format("Dequeued latency info [%s ms]: %s", elapsed, latencyDetails));
```

Comment:
any reason to not use {} and SLF4J's own String expansion?

Reply:
It was a miss from my side.